src/evm.rs

```diff
@@ -1,179 +1,338 @@
+    }
 
-        Ok(EVM(executor))
+    /// Get storage value of address at index.
+    fn storage(&mut self, address: &str, index: U256) -> PyResult<Option<U256>> {
+        let (account, _) = self.context.load_account(addr(address)?).map_err(pyerr)?;
+        Ok(account.storage.get(&index).map(|s| s.present_value))
     }
 
-    /// Inserts the provided account information in the database at
-    /// the specified address.
-    fn basic(mut _self: PyRefMut<'_, Self>, address: &str) -> PyResult<Option<AccountInfo>> {
-        let db = &mut _self.0.backend;
-        let acc = db.basic(addr(address)?).map_err(pyerr)?;
-        Ok(acc.map(Into::into))
+    /// Get block hash by block number.
+    fn block_hash(&mut self, number: U256, py: Python<'_>) -> PyResult<Option<PyObject>> {
+        let bytes = self.context.block_hash(number).map_err(pyerr)?;
+        if bytes.is_empty() {
+            return Ok(None);
+        }
+        Ok(Some(PyBytes::new(py, &bytes.to_vec()).into()))
     }
 
-    /// Inserts the provided account information in the database at
-    /// the specified address.
+    /// Inserts the provided account information in the database at the specified address.
     fn insert_account_info(
-        mut _self: PyRefMut<'_, Self>,
+        &mut self,
         address: &str,
         info: AccountInfo,
     ) -> PyResult<()> {
-        let db = &mut _self.0.backend;
-        db.insert_account_info(addr(address)?, info.into());
-
+        let target = addr(address)?;
+        match self.context.journaled_state.state.get_mut(&target) {
+            // account is cold, just insert into the DB, so it's retrieved next time
+            None => self.context.db.insert_account_info(target, info.into()),
+            // just replace the account info
+            Some(acc) => acc.info = info.into(),
+        }
         Ok(())
     }
 
     /// Set the balance of a given address.
-    fn set_balance(mut _self: PyRefMut<'_, Self>, address: &str, balance: U256) -> PyResult<()> {
-        _self
-            .0
-            .set_balance(addr(address)?, balance)
-            .map_err(pyerr)?;
+    fn set_balance(&mut self, address: &str, balance: U256) -> PyResult<()> {
+        let address_ = addr(address)?;
+        let account = {
+            let (account, _) = self.context.load_account(address_).map_err(pyerr)?;
+            account.info.balance = balance;
+            account.clone()
+        };
+        self.context.journaled_state.state.insert(address_, account);
+        self.context.journaled_state.touch(&address_);
         Ok(())
     }
 
     /// Retrieve the balance of a given address.
-    fn get_balance(_self: PyRef<'_, Self>, address: &str) -> PyResult<U256> {
-        let balance = _self.0.get_balance(addr(address)?).map_err(pyerr)?;
+    fn get_balance(&mut self, address: &str) -> PyResult<U256> {
+        let (balance, _) = self.context.balance(addr(address)?).map_err(pyerr)?;
```

Comment:
Are we okay to Error on account that are inside Evm?

Reply:
Why is this going to error? The `context.balance` function seems to look first in the journaled state and then the db :thinking: